### PR TITLE
Improve the parsing of methods in MySQL

### LINKED ISSUE #31558: Improve the parsing of methods in MySQL
## Background

Hi community.\
The ShardingSphere SQL parser engine helps users to parse SQL to create the AST (Abstract Syntax Tree) and visit the AST to get SQLStatement (Java Object). 

Currently, we are planning to enhance the support for MySQL SQL parsing in ShardingSphere.

More details:
https://shardingsphere.apache.org/document/current/en/reference/sharding/parse/

## Issue Background Explanation
In the original parsing work, it may have overlooked the parsing of method parameters. For ShardingSphere, we need to pay attention to the table name or field name in method parameters. If there are issues with parsing method parameters, it will cause problems in subsequent binding and rewriting tasks.

For the verification work, we need to complete the following items.
1. Find the example SQL of this method in the official website.[Built-In Function and Operator Reference](https://dev.mysql.com/doc/refman/8.0/en/built-in-function-reference.html)
2. verify if the SQL itself can be parsed by the parser
3. Check if the parsed SQLStatement can correctly capture the parameters in the method.

## Task

```
MD5()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/encryption-functions.html#function_md5)

```
MEMBER OF()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#operator_member-of)
```
MICROSECOND()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_microsecond)
```
MID()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_mid)
```
MIN()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions.html#function_min)
```
MINUTE()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_minute)
```
MOD()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/mathematical-functions.html#function_mod)
```
MONTH()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_month)
```
MONTHNAME()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_monthname)
```
MultiLineString()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/gis-mysql-specific-functions.html#function_multilinestring)
```
MultiPoint()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/gis-mysql-specific-functions.html#function_multipoint)
```
MultiPolygon()
```
- [ ] [link](https://dev.mysql.com/doc/refman/8.0/en/gis-mysql-specific-functions.html#function_multipolygon)

## Overall Procedure

If you intend to participate in fixing this issue, please feel free to leave a comment below the issue. Community members will assign the issue accordingly.  

><u>For example, you can leave a comment like this: "Hi, please assign this issue to me. Thank you!"</u>

Once you have claimed the issue, please review the syntax of the SQL on the official website of the corresponding database. Execute the SQL on the respective database to ensure the correctness of the SQL syntax.

><u>You can check the corresponding source of each SQL case on the official database website by clicking on the link provided below each case.</u> 

Next, execute the problematic SQL cases mentioned above in the database (you can quickly start the                     corresponding database using the Docker image for that database, and then connect to it using 
a client you are  familiar with), to ensure that the SQL syntax itself is correct.  

## Fixing ANTLR Grammar Parsing Issue

Once you have confirmed the correctness of the SQL syntax, you can validate and fix the grammar parsing issue in ShardingSphere. 

>If you are using IntelliJ IDEA, you will need to install the ANTLR plugin before proceeding.

If it is an ANTLR parsing error message, try to repair the `.g4` file by comparing it with the official database syntax until the SQL can be correctly parsed by ANTLR.

When there is no error message in the ANTLR Preview window, it means that ANTLR can correctly parse the SQL.  

## Visitor problem fix

After ANTLR parses SQL into an abstract syntax tree, ShardingSphere will access the abstract syntax tree through Visitor and extract the required information.
If you need to extract Segments, you need to first execute:
 ```
 mvn -T 2C clean install -DskipTests
 ```  
 Under the shardingsphere-parser module to compile the entire parser module.\
Then rewrite the corresponding visit method in SQLStatementVisitorr as needed to extract the corresponding Segment.

## Add assertion test file  

After the above SQL parsing problem is repaired, the corresponding Test needs to be added. 
The steps are as follows:
1. Add the corresponding `sql-case` in the `sql/supported` directory.
2. Add case assertions in the case directory of the `shardingsphere-test-it-parser` module.
3.  Run `org.apache.shardingsphere.test.it.sql.parser.internal.InternalSQLParserIT`\
 After SQL Parser IT runs successfully, you can submit a PR.   

## Relevant Skills

1. Master JAVA language
2. Have a basic understanding of Antlr g4 file
3. Be familiar with Doris SQLs

### PR DESCRIPTION
**Pull request by lushaorong**
Fixes #31558.

Changes proposed in this pull request:
  - Improve the parsing of methods in MySQL

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [x] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: test/it/parser/src/main/resources/sql/supported/dml/select-special-function.xml [hunk 1]
```diff
@@ -180,4 +180,16 @@
...
+    <sql-case id="select_microsecond" value="SELECT MICROSECOND('12:00:00.123456')" db-types="MySQL" />
+    <sql-case id="select_mid" value="SELECT MID('foobarbar' from 4)" db-types="MySQL" />
+    <sql-case id="select_minute" value="SELECT MINUTE('2008-02-03 10:05:03')" db-types="MySQL" />
+    <sql-case id="select_mod" value="SELECT MOD(234, 10)" db-types="MySQL" />
+    <sql-case id="select_month" value="SELECT MONTH('2008-02-03')" db-types="MySQL" />
+    <sql-case id="select_monthname" value="SELECT MONTHNAME('2008-02-03')" db-types="MySQL" />
+    <sql-case id="select_multilinestring" value="SELECT ST_GeomFromText('MultiLineString((0 0, 1 1, 2 2),(3 3, 4 4, 5 5))')" db-types="MySQL" />
+    <sql-case id="select_multipoint" value="SELECT ST_GeomFromText('MultiPoint((1 2), (3 4), (5 6))')" db-types="MySQL" />
+    <sql-case id="select_multipolygon" value="SELECT ST_GeomFromText('MultiPolygon(((0 0, 1 0, 1 1, 0 1, 0 0)),((2 2, 3 2, 3 3, 2 3, 2 2)))')" db-types="MySQL" />
+
```
Comment (strongduanmu) [inline test/it/parser/src/main/resources/sql/supported/dml/select-special-function.xml:194]: Please remove useless blank line.

File: test/it/parser/src/main/resources/case/dml/select-special-function.xml [hunk 2]
```diff
@@ -3020,4 +3020,168 @@
...
+        </projections>
+    </select>
+
+    <select sql-case-id="select_multilinestring">
+        <projections start-index="7" stop-index="73">
+            <expression-projection start-index="7" stop-index="73" text="ST_GeomFromText('MultiLineString((0 0, 1 1, 2 2),(3 3, 4 4, 5 5))')">
+                <expr>
+                    <function function-name="ST_GeomFromText" text="ST_GeomFromText('MultiLineString((0 0, 1 1, 2 2),(3 3, 4 4, 5 5))')" start-index="7" stop-index="73">
+                        <parameter>
+                            <literal-expression value="MultiLineString((0 0, 1 1, 2 2),(3 3, 4 4, 5 5))" start-index="23" stop-index="72" />
```
Comment (strongduanmu) [inline test/it/parser/src/main/resources/case/dml/select-special-function.xml:3152]: Should MultiLineString also be parsed as a function?

File: test/it/parser/src/main/resources/case/dml/select-special-function.xml [hunk 3]
```diff
@@ -3020,4 +3020,168 @@
...
+        </projections>
+    </select>
+
+    <select sql-case-id="select_multipoint">
+        <projections start-index="7" stop-index="56">
+            <expression-projection start-index="7" stop-index="56" text="ST_GeomFromText('MultiPoint((1 2), (3 4), (5 6))')">
+                <expr>
+                    <function function-name="ST_GeomFromText" text="ST_GeomFromText('MultiPoint((1 2), (3 4), (5 6))')" start-index="7" stop-index="56">
+                        <parameter>
+                            <literal-expression value="MultiPoint((1 2), (3 4), (5 6))" start-index="23" stop-index="55" />
```
Comment (strongduanmu) [inline test/it/parser/src/main/resources/case/dml/select-special-function.xml:3166]: Should MultiPoint also be parsed as a function?

File: test/it/parser/src/main/resources/case/dml/select-special-function.xml [hunk 4]
```diff
@@ -3020,4 +3020,168 @@
...
+        </projections>
+    </select>
+
+    <select sql-case-id="select_multipolygon">
+        <projections start-index="7" stop-index="94">
+            <expression-projection start-index="7" stop-index="94" text="ST_GeomFromText('MultiPolygon(((0 0, 1 0, 1 1, 0 1, 0 0)),((2 2, 3 2, 3 3, 2 3, 2 2)))')">
+                <expr>
+                    <function function-name="ST_GeomFromText" text="ST_GeomFromText('MultiPolygon(((0 0, 1 0, 1 1, 0 1, 0 0)),((2 2, 3 2, 3 3, 2 3, 2 2)))')" start-index="7" stop-index="94">
+                        <parameter>
+                            <literal-expression value="MultiPolygon(((0 0, 1 0, 1 1, 0 1, 0 0)),((2 2, 3 2, 3 3, 2 3, 2 2)))" start-index="23" stop-index="93" />
```
Comment (strongduanmu) [inline test/it/parser/src/main/resources/case/dml/select-special-function.xml:3180]: Should MultiPolygon also be parsed as a function?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: test/it/parser/src/main/resources/case/dml/select-special-function.xml (hunk 2 above)
Comment (lushaorong) [reply test/it/parser/src/main/resources/case/dml/select-special-function.xml:3152]: > Should MultiLineString also be parsed as a function?

`MultiLineString((0 0, 1 1, 2 2),(3 3, 4 4, 5 5))` is the string parameter of st_geomfromtext. How should I proceed with this?

Comment (lushaorong) [reply test/it/parser/src/main/resources/case/dml/select-special-function.xml:3152]: image

image
`(0 0, 1 1, 2 2)`  cannot be parse to `literal-expression`

Comment (strongduanmu) [reply test/it/parser/src/main/resources/case/dml/select-special-function.xml:3152]: > image
> 
> image `(0 0, 1 1, 2 2)` cannot be parse to `literal-expression`

It looks great.

**Review by strongduanmu — APPROVED** [review]
Looks so great, merged.

**Comment by strongduanmu** [comment]
Hi @lushaorong, could you please check if Doris supports these statements? If so, please submit another PR.

**Comment by lushaorong** [comment]
> Hi @lushaorong, could you please check if Doris supports these statements? If so, please submit another PR.

I can try it

**Pull request merged**